HeuristicNew.java

```diff
@@ -0,0 +1,256 @@
+public class HeuristicNew {
+	public static final int NUM_FEATURES = 6;
+	static double[] weights;
+	static int weightCounter;
+
+	public static double evaluate(TempState s, double[] inputWeights) {
+		weightCounter = 0;
+		readWeights();
+		int[][] field = s.getField();
+
+        if (inputWeights == null) {
+            weights = getDefaultWeights();
+        } else {
+            weights = inputWeights;
+        }
+
+        int[] colHeights = colHeights(field);
+        return weightedLandingHeight(s) + weightedRowsEliminated(s.getRowsCleared()) + weightedNumRowTranstions(field) + weightedNumColTranstions(field)
+                + wNumHoles(field, colHeights) + weightedWellSums(field);
+	}
+
+    public static double evaluate(TempState s) {
+        return evaluate(s, getDefaultWeights());
+    }
+
+    // TODO: either remove or read from file
+    private static double[] getDefaultWeights() {
+        weights = new double[NUM_FEATURES];
+        for (int i = 0; i < weights.length; i++) {
+            weights[i] = 1;
+        }
+        return weights;
+    }
+
+    // TODO: read from file instead
+    private static void readWeights() {
+        weights = new double[NUM_FEATURES];
+        for (int i = 0; i < weights.length; i++) {
+            weights[i] = 1;
+        }
+    }
+
+    private static double weightedLandingHeight (TempState s) {
+	    return getNextWeight() * landingHeight(s);
+    }
+
+    // Feature 1
+    // Landing Height:
+    // The height where the piece is put = the height of the column BEFORE piece is put + (the height of the piece / 2)
+    // Also equivalent to height of column AFTER piece is put - (the height of the piece / 2) (?)
+    private static int landingHeight(TempState s) {
+
+	    int heightPiece = s.getHeightOfPeice();
+	    int landingHeight = s.getHeightOfCol(s.getStateSlot()) - heightPiece/2;
+
+	    return landingHeight;
+    }
+
+    // Feature 2
+    private static double weightedRowsEliminated (int rowsCleared) {
+	    return getNextWeight() * rowsCleared;
+    }
+
+    private static double weightedNumRowTranstions(int[][] field) {
+        return getNextWeight() * numRowTransitions(field);
+    }
+
+    // Feature 3
+    // The total number of row transitions.
+    // A row transition occurs when an empty cell is adjacent to a filled cell
+    // on the same row and vice versa.
+	private static int numRowTransitions(int[][] field) {
+	    int count = 0;
+	    int numRows = field.length;
+	    int numCols = field[0].length;
+        for (int i = 0; i < numRows; i++) {
+            for (int j = 0; j < numCols - 1; j++) {
+                if (field[i][j] == 0 && field[i][j+1] != 0
```

Comment:
won't this cause double counting?

e.g. 0101 - output will be 3 instead of 2

Reply:
It is correct. The answer is 3.

http://imake.ninja/el-tetris-an-improvement-on-pierre-dellacheries-algorithm/
From here, the author explains himself:

> X_X_X_X_X (If you read this row from left to right, we keep crossing from X to _ and then from _ to X. Therefore, there are 8 row transitions

Reply:
i see!